src/Microsoft.Health.Dicom.Core/Features/Indexing/IInstanceReindexer.cs

```diff
@@ -0,0 +1,27 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Health.Dicom.Core.Features.ExtendedQueryTag;
+
+namespace Microsoft.Health.Dicom.Core.Features.Indexing
+{
+    /// <summary>
+    /// Reindexer which reindexes DICOM instance.
+    /// </summary>
+    public interface IInstanceReindexer
+    {
+        /// <summary>
+        /// Reindex DICOM instance of watermark on extended query tags.
+        /// </summary>
+        /// <param name="entries">Extended query tag store entries.</param>
+        /// <param name="watermark">The watermark to DICOM instance.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The task.</returns>
+        Task ReindexInstanceAsync(IEnumerable<ExtendedQueryTagStoreEntry> entries, long watermark, CancellationToken cancellationToken = default);
```

Comment:
nit: I think we can be a bit more restrictive and ask for `ICollection<ExtendedQueryTagStoreEntry>`

Reply:
Changed to IReadonlyList